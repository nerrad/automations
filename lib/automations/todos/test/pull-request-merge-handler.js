const {
  gimmeOctokit,
  gimmeContext,
  gimmeApp,
  loadDiff,
} = require("@testHelpers");
const { reset } = require("../utils/check-for-duplicate-issue");

describe("pull-request-merged-handler", () => {
  let octokit, context, app;
  beforeEach(() => {
    octokit = gimmeOctokit();
    context = gimmeContext("pull_request", "closed");
    app = gimmeApp("todos");
    reset();
  });
  it("does nothing on an unmerged, closed PR", async () => {
    context.payload = { ...context.payload, pull_request: { merged: false } };
    await app(context, octokit);
    expect(octokit.issues.create).not.toHaveBeenCalled();
  });
  it("creates an issue", async () => {
    await app(context, octokit);
    expect(octokit.issues.create).toHaveBeenCalledTimes(1);
    expect(octokit.issues.create.mock.calls[0]).toMatchSnapshot();
  });
  it("creates an issue with a truncated title", async () => {
    octokit.pulls.get.mockReturnValue(loadDiff("long-title"));
    await app(context, octokit);
    expect(octokit.issues.create).toHaveBeenCalledTimes(1);
    expect(octokit.issues.create.mock.calls[0]).toMatchSnapshot();
  });
  it("does not create any issues if no todos are found", async () => {
    octokit.pulls.get.mockReturnValue(loadDiff("none"));
    await app(context, octokit);
    expect(octokit.issues.create).not.toHaveBeenCalled();
  });
  it("does not create an issue that already exists", async () => {
    octokit.search.issuesAndPullRequests.mockReturnValueOnce(
      Promise.resolve({
        data: {
          total_count: 1,
          items: [{ title: "I am an example title", state: "open" }],
        },
      })
    );
    await app(context, octokit);
    expect(octokit.issues.create).not.toHaveBeenCalled();
  });
  it("creates many (5) issues", async () => {
    octokit.pulls.get.mockReturnValue(loadDiff("many"));
    await app(context, octokit);
    expect(octokit.issues.create).toHaveBeenCalledTimes(5);
  });
  it("reopens a closed issue", async () => {
    octokit.search.issuesAndPullRequests.mockReturnValueOnce(
      Promise.resolve({
        data: {
          total_count: 1,
          items: [
            { number: 1, title: "I am an example title", state: "closed" },
          ],
        },
      })
    );
    await app(context, octokit);
    expect(octokit.issues.update).toHaveBeenCalledTimes(1);
    expect(octokit.issues.createComment).toHaveBeenCalledTimes(1);
    expect(octokit.issues.createComment.mock.calls[0]).toMatchSnapshot();
    expect(octokit.issues.create).not.toHaveBeenCalled();
  });
});
