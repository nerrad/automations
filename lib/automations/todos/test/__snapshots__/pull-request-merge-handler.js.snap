// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pull-request-merged-handler creates an issue 1`] = `
Array [
  Object {
    "body": "I am an example title

---
https://github.com/nerrad/tests/blob/6c2373ad036d292f7e9467981ba8b467e8b8b0ff/.travis.yml#L9-L15
---
###### :rocket: This issue was generated by the [automations bot](https://github.com/nerrad/automations) based on a \`todo\` comment in 6c2373ad036d292f7e9467981ba8b467e8b8b0ff when #21 was merged. cc @nerrad",
    "owner": "nerrad",
    "repo": "tests",
    "title": "I am an example title",
  },
]
`;

exports[`pull-request-merged-handler creates an issue with a truncated title 1`] = `
Array [
  Object {
    "body": "This is a really, really, really, really, really, really, really, really, really, really long title

---
https://github.com/nerrad/tests/blob/6c2373ad036d292f7e9467981ba8b467e8b8b0ff/.travis.yml#L9-L15
---
###### :rocket: This issue was generated by the [automations bot](https://github.com/nerrad/automations) based on a \`todo\` comment in 6c2373ad036d292f7e9467981ba8b467e8b8b0ff when #21 was merged. cc @nerrad",
    "owner": "nerrad",
    "repo": "tests",
    "title": "This is a really, really, really, really, really, really,...",
  },
]
`;

exports[`pull-request-merged-handler reopens a closed issue 1`] = `
Array [
  Object {
    "body": ":rocket: This issue has been reopened because the **\`todo\`** comment still exists in [****](https://github.com/nerrad/tests/blob/6c2373ad036d292f7e9467981ba8b467e8b8b0ff/), as of 6c2373ad036d292f7e9467981ba8b467e8b8b0ff.

---
###### If this was not intentional, just remove the comment from your code.",
    "issue_number": 1,
    "owner": "nerrad",
    "repo": "tests",
  },
]
`;
